vpr/src/pack/greedy_seed_selector.cpp

```diff
@@ -0,0 +1,237 @@
+/**
+ * @file
+ * @author  Alex Singer
+ * @date    November 2024
+ * @brief   The definitions of the Greedy Seed Selector class.
+ */
+
+#include "greedy_seed_selector.h"
+
+#include <algorithm>
+#include "atom_netlist.h"
+#include "cluster_legalizer.h"
+#include "cluster_util.h"
+#include "echo_files.h"
+#include "prepack.h"
+#include "vpr_error.h"
+#include "vpr_types.h"
+#include "vtr_assert.h"
+#include "vtr_math.h"
+#include "vtr_vector.h"
+
+/**
+ * @brief Helper method that computes the seed gain of the give atom block.
+ *
+ * The seed_type variable selects which algorithm to use to compute the seed
+ * gain.
+ */
+static inline float get_seed_gain(AtomBlockId blk_id,
+                                  const AtomNetlist& atom_netlist,
+                                  const Prepacker& prepacker,
+                                  const e_cluster_seed seed_type,
+                                  const t_molecule_stats& max_molecule_stats,
+                                  const vtr::vector<AtomBlockId, float>& atom_criticality) {
+    switch (seed_type) {
+        // By criticality
+        case e_cluster_seed::TIMING:
+            return atom_criticality[blk_id];
+        // By number of used molecule input pins
+        case e_cluster_seed::MAX_INPUTS:
+        {
+            const t_pack_molecule* blk_mol = prepacker.get_atom_molecule(blk_id);
+            const t_molecule_stats molecule_stats = calc_molecule_stats(blk_mol, atom_netlist);
+            return molecule_stats.num_used_ext_inputs;
+        }
+        // By blended gain (criticality and inputs used)
+        case e_cluster_seed::BLEND:
+        {
+            // Score seed gain of each block as a weighted sum of timing
+            // criticality, number of tightly coupled blocks connected to
+            // it, and number of external inputs.
+            float seed_blend_fac = 0.5f;
+
+            const t_pack_molecule* blk_mol = prepacker.get_atom_molecule(blk_id);
+            const t_molecule_stats molecule_stats = calc_molecule_stats(blk_mol, atom_netlist);
+            VTR_ASSERT(max_molecule_stats.num_used_ext_inputs > 0);
+
+            float used_ext_input_pin_ratio = vtr::safe_ratio<float>(molecule_stats.num_used_ext_inputs, max_molecule_stats.num_used_ext_inputs);
+            float blend_gain = (seed_blend_fac * atom_criticality[blk_id]
+                                + (1 - seed_blend_fac) * used_ext_input_pin_ratio);
+            blend_gain *= (1 + 0.2 * (molecule_stats.num_blocks - 1));
+            return blend_gain;
+        }
+        // By pins per molecule (i.e. available pins on primitives, not pins in use)
```

Comment:
Intuition (a weak one): primitive types with more pins or inputs might be harder to pack.